### PR TITLE
Scrubber rework, CPDShell confusion matrix and other metrics computation

### PR DESCRIPTION
**Pull request by wrdxwrdxwrdx**
An abstract scrubber has been added and the management of shared data has been transferred to the Data Controller.

### CODE REVIEW

File: CPDShell/Core/cpd_core.py [hunk 1]
```diff
@@ -1,33 +1,56 @@
...
         self.scrubber = scrubber
+        self.scenario = scenario
         self.algorithm = algorithm
 
     def run(self) -> list[int]:
         """Find change points
 
         :return: list of change points
         """
+        self.scrubber.scenario = self.scenario
```
Comment (vkutuev) [inline CPDShell/Core/cpd_core.py:43]: Maybe move it to `__init__` method?

File: CPDShell/Core/data_controller.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+from collections.abc import Iterable
```
Comment (vkutuev) [inline CPDShell/Core/data_controller.py:9]: Add copyright here like [this](https://github.com/Lesh79/PySATL-CPD-Module/blob/main/CPDShell/Core/algorithms/KNNCPD/knn_graph.py#L1)

File: CPDShell/Core/scrubber/abstract_scrubber.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        """Function for restarting Scrubber"""
+        ...
+
+    @abstractmethod
+    def get_windows(self) -> Iterable[Sequence[float | numpy.float64]]:
+        """Function for dividing data into parts to feed into the change point detection algorithm
+
+        :return: Iterator of data windows for change point detection algorithm
+        """
+        ...
```
Comment (vkutuev) [inline CPDShell/Core/scrubber/abstract_scrubber.py]:
```suggestion
        raise NotImplementedError
```

File: CPDShell/Core/scrubber/abstract_scrubber.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        """
+        self._scenario: Scenario | None = None
+        self._data: Sequence[float | numpy.float64] = []
+        self.is_running = True
+        self.change_points: list[int] = []
+
+    @abstractmethod
+    def restart(self) -> None:
+        """Function for restarting Scrubber"""
+        ...
```
Comment (vkutuev) [inline CPDShell/Core/scrubber/abstract_scrubber.py]:
```suggestion
        raise NotImplementedError
```

File: CPDShell/Core/scrubber/abstract_scrubber.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        """Function for dividing data into parts to feed into the change point detection algorithm
+
+        :return: Iterator of data windows for change point detection algorithm
+        """
+        ...
+
+    @abstractmethod
+    def add_change_points(self, window_change_points: list[int]) -> None:
+        """Function for mapping window change points to scrubber data part"""
+        ...
```
Comment (vkutuev) [inline CPDShell/Core/scrubber/abstract_scrubber.py]:
```suggestion
        raise NotImplementedError
```

File: CPDShell/Core/scrubber/abstract_scrubber.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
+from abc import ABC, abstractmethod
```
Comment (vkutuev) [inline CPDShell/Core/scrubber/abstract_scrubber.py:9]: Add copyright

File: CPDShell/Core/scrubber/linear_scrubber.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+from collections.abc import Iterable, Sequence
```
Comment (vkutuev) [inline CPDShell/Core/scrubber/linear_scrubber.py:9]: Add copyright

File: tests/test_CPDShell/test_Core/test_data_controller.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
+import pytest
```
Comment (vkutuev) [inline tests/test_CPDShell/test_Core/test_data_controller.py:12]: Add copyright

**Review by vkutuev — CHANGES_REQUESTED** [review]

File: CPDShell/Core/scrubber/linear_scrubber.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    def restart(self) -> None:
+        self.change_points = []
+        self.is_running = True
+
+    def get_windows(self) -> Iterable[Sequence[float | numpy.float64]]:
+        if self._data:
+            window_end = self._window_start + self._window_length
+            yield self._data[self._window_start : window_end]
+            self._window_start += int(self._window_length * self._movement_k)
+        while self._window_start + self._window_length <= len(self._data) and self.is_running:
```
Comment (artemiipatov) [inline CPDShell/Core/scrubber/linear_scrubber.py]:
```suggestion
        while (self._data and self._window_start == 0
               or self._window_start + self._window_length <= len(self._data) and self.is_running):
```

File: CPDShell/Core/scrubber/linear_scrubber.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            window_end = self._window_start + self._window_length
+            yield self._data[self._window_start : window_end]
+            self._window_start += int(self._window_length * self._movement_k)
+
+    def add_change_points(self, window_change_points: list[int]) -> None:
```
Comment (artemiipatov) [inline CPDShell/Core/scrubber/linear_scrubber.py:51]: There was a little optimization in the previous version: if we detect some change point, we do not include it and the preceding points in the next window. I'm not sure if it is good or not. But it seems, that in this version the same points can be included twice.

File: CPDShell/Core/scrubber/abstract_scrubber.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+from collections.abc import Iterable, Sequence
+
+import numpy
+
+from CPDShell.Core.scenario import Scenario
+
+
+class AbstractScrubber(ABC):
```
Comment (artemiipatov) [inline CPDShell/Core/scrubber/abstract_scrubber.py]: Probably it would be better to rename it to just `Scrubber`?

File: CPDShell/shell.py [hunk 12]
```diff
@@ -97,26 +98,25 @@ class CPDShell:
...
-        scrubber_class = scrubber_class if scrubber_class is not None else Scrubber
         arg = 5
         cpd_algorithm = (
             cpd_algorithm if cpd_algorithm is not None else GraphAlgorithm(lambda a, b: abs(a - b) <= arg, 2)
         )
         self.cpd_core: CPDCore = CPDCore(
-            scrubber_class(Scenario(10, True), data.raw_data if isinstance(data, LabeledCPData) else data),
-            cpd_algorithm,
-        )  # if no algo or scrubber was given, then some standard
+            scenario, data.raw_data if isinstance(data, LabeledCPData) else data, scrubber, cpd_algorithm, 10
```
Comment (artemiipatov) [inline CPDShell/shell.py]: Strange default value of `scrubber_data_size`. It is less than the window size in almost all cases.

File: CPDShell/Core/scrubber/abstract_scrubber.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    @abstractmethod
+    def get_windows(self) -> Iterable[Sequence[float | numpy.float64]]:
+        """Function for dividing data into parts to feed into the change point detection algorithm
+
+        :return: Iterator of data windows for change point detection algorithm
+        """
+        ...
+
+    @abstractmethod
+    def add_change_points(self, window_change_points: list[int]) -> None:
```
Comment (artemiipatov) [inline CPDShell/Core/scrubber/abstract_scrubber.py:48]: What about documentation for parameters?

File: CPDShell/Core/scrubber/abstract_scrubber.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    by change point detection algorithms
+    """
+
+    def __init__(self) -> None:
+        """A scrubber for dividing data into windows
+        and subsequent processing of data windows
+        by change point detection algorithms
+
+        """
+        self._scenario: Scenario | None = None
```
Comment (artemiipatov) [inline CPDShell/Core/scrubber/abstract_scrubber.py]: Should we include `Scenario` type in alpha version? We don't have any other scrubbers for now. It overcomplicates algorithm construction, making one more layer.

**Review by artemiipatov — CHANGES_REQUESTED** [review]

File: CPDShell/Core/scrubber/linear_scrubber.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+    def restart(self) -> None:
+        self.change_points = []
+        self.is_running = True
+
+    def get_windows(self) -> Iterable[Sequence[float | numpy.float64]]:
+        if self._data:
+            window_end = self._window_start + self._window_length
+            yield self._data[self._window_start : window_end]
+            self._window_start += int(self._window_length * self._movement_k)
```
Comment (artemiipatov) [inline CPDShell/Core/scrubber/linear_scrubber.py]: Now you can get rid of this `if`.

File: CPDShell/Core/data_controller.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        """Function for getting data pieces
+
+        :return: data pieces iterator"""
+        while self._data_start_index < len(self._data):
+            cur_index = self._data_start_index
+            yield self._data[cur_index : self._data_start_index + self._scrubber_data_size]
+            self._data_start_index += self._scrubber_data_size
+
+    def add_change_points(self, window_change_points) -> None:
+        """Function for mapping window change points to global data"""
```
Comment (artemiipatov) [inline CPDShell/Core/data_controller.py]: No parameter's documentation.

File: CPDShell/Core/scrubber/abstract_scrubber.py [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+    @abstractmethod
+    def add_change_points(self, window_change_points: list[int]) -> None:
+        """Function for mapping window change points to scrubber data part
+        :param window_change_points: change points detected on scrubber window
+        """
+        raise NotImplementedError
+
+    def add_data(self, new_data: Sequence[float | numpy.float64]) -> None:
+        """Function for adding new data to Scrubber"""
```
Comment (artemiipatov) [inline CPDShell/Core/scrubber/abstract_scrubber.py]: No doc.

**Review by artemiipatov — CHANGES_REQUESTED** [review]

**Review by artemiipatov — APPROVED** [review]

File: CPDShell/Core/cpd_core.py [hunk 18]
```diff
@@ -1,33 +1,56 @@
...
-            if self.scrubber.scenario.to_localize:
-                window_change_points = self.algorithm.localize(window)
-            else:
-                change_points_number = self.algorithm.detect(window)
-                window_change_points = [self.scrubber.window_length] * change_points_number
-            self.scrubber.add_change_points(window_change_points)
-        return self.scrubber.change_points
+        self.data_controller.restart()
+        for data in self.data_controller.get_data():
+            self.scrubber._data = data
```
Comment (artemiipatov) [inline CPDShell/Core/cpd_core.py]: Restart scrubber.

**Review by artemiipatov — APPROVED** [review]

**Review by artemiipatov — APPROVED** [review]

**Review by vkutuev — APPROVED** [review]

**Pull request merged**